middleware/src/middleware.go

```diff
@@ -335,56 +381,211 @@ func (middleware *Middleware) RestoreHSMSecret() rpcmessages.ErrorResponse {
+			},
+		}
+	}
+
+	passwordFromStorage := usersMap[args.Username].BCryptedPassword
+	err = bcrypt.CompareHashAndPassword([]byte(passwordFromStorage), []byte(args.Password))
+	if err != nil {
+		log.Println("Hash and password did not match")
+		return rpcmessages.UserAuthenticateResponse{
+			ErrorResponse: &rpcmessages.ErrorResponse{
+				Success: false,
+				Message: "authentication unsuccessful, incorrect password",
+				Code:    rpcmessages.ErrorAuthenticationPasswordIncorrect,
+			},
+		}
+	}
+
+	jwtTokenStr, err := middleware.jwtAuth.GenerateToken(args.Username)
+	if err != nil {
+		return rpcmessages.UserAuthenticateResponse{
+			ErrorResponse: &rpcmessages.ErrorResponse{
+				Success: false,
+				Message: "authentication unsuccessful, jwt error",
+				Code:    rpcmessages.ErrorAuthenticationFailed,
+			},
+		}
+	}
+
+	return rpcmessages.UserAuthenticateResponse{
+		ErrorResponse: &rpcmessages.ErrorResponse{
+			Success: true,
+		},
+		Token: jwtTokenStr,
 	}
 }
 
 // UserChangePassword returns an ErrorResponse struct in response to a rpcserver request
-// FIXME: This is a dummy implementation of the UserChangePassword RPC call
-// This dummy method approves all passwords which are longer or equal to 8 chars
+// The function first validates the current password with redis, then replaces it with the new password.
+// Passwords need to be longer than or equal to 8 chars.
 func (middleware *Middleware) UserChangePassword(args rpcmessages.UserChangePasswordArgs) rpcmessages.ErrorResponse {
-	if len(args.NewPassword) >= 8 {
-		if args.Username == "admin" {
-			middleware.dummyAdminPassword = args.NewPassword
-			if !middleware.dummyIsBaseSetup {
-				middleware.dummyIsBaseSetup = true // the change of the admin password completes the dummy setup process (for now)
-			}
+	if len(args.NewPassword) < 8 {
+		return rpcmessages.ErrorResponse{
+			Success: false,
+			Message: "password change unsuccessful, the password needs to be at least 8 characters in length",
+			Code:    rpcmessages.ErrorPasswordTooShort,
 		}
-		return rpcmessages.ErrorResponse{Success: true}
 	}
 
-	return rpcmessages.ErrorResponse{
-		Success: false,
-		Message: "password change unsuccessful (too short)",
-		Code:    rpcmessages.ErrorDummyPasswordTooShort,
+	usersMap := make(map[string]UserAuthStruct)
+	authStructureString, err := middleware.redisClient.GetString(redis.MiddlewareAuth)
+	if err != nil {
+		log.Println("error when getting the authentication structure from redis")
+		return rpcmessages.ErrorResponse{
+			Success: false,
+			Message: "password change unsuccessful",
+			Code:    rpcmessages.ErrorPasswordChangeFailed,
+		}
+	}
+	err = json.Unmarshal([]byte(authStructureString), &usersMap)
+	if err != nil {
+		log.Println("Did not receive json from redis's authentication structure")
+		return rpcmessages.ErrorResponse{
+			Success: false,
+			Message: "password change unsuccessful",
+			Code:    rpcmessages.ErrorPasswordChangeFailed,
+		}
+	}
```

Comment:
You use pretty much the same block of code with `authStructureString, err := middleware.redisClient.GetString(redis.MiddlewareAuth)` and `err = json.Unmarshal([]byte(authStructureString), &usersMap)` twice. Once in `UserAuthenticate` and again here in `UserChangePassword`. Even the error messages have slightly different wording. e.g. `error getting the auth structure string from the redis client` vs. `error when getting the authentication structure from redis`

Maybe refactor this into a util function, something like `getAuthStructure` or `handleAuthStructure` that handles both the getting and the unmarshal?

Reply:
Great, I moved it to a separate util function.